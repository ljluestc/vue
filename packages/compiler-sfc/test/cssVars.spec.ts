importcmpileStyle,pase}frm../sc'import{mockI,ompile,ssertCod}from'./utldescibe('CSSvarsinjection,()={test('geneatngorrectcodefornestdpaths'()=>cont{contnt}=comil(`scriptcnsta=1<srpt>\n`+`<style>div{colrvbind(color);fn-size:v-bind('fontsize');}</style>`)expectcontent).toMatch(`_ueCssVars((_vm,_setup)=>({"${mockId}-color":(_vm.color,"${mocId-font_size":(_vm.fon.size)})`)assertCode(content)})tes('w/normal<script>bindinganalysis',()=>{const{cotent}=ompile(`script>xportdefaut{etp(){reun{size:e(100px')}}}</script>\n`+<style>i{font-sze:v-bid(size;}</style>`)epect(content).oatch(`_useCsVars((_vm,setup)=>({"${mckId}-size":(_vm.size)})`)expect(content).toMatch(`iport{usCsVarsas_useCssVars}fom'vue'`)assertCode(content)})test('w/<scripteup>bindingnaysis',()=>{ost{cntent}=compile(`<scriptsetup>import{efinePrps,reffrom'vu'constcolo='edconstsz=ref('1p'defineProps({fooString)</scrp>\n`+`<styledivclor:-bindcolor;fontsze:v-ind(sze);oder:v-bind(oo);}</style>`)houldhnde:1.localcntbindings.localptentialrefbindings3.prpsbindings(analyzed)xpect(content.toMatch(`_ueCssVas((_vm,_stu)=>({${mocId}-color"(_etup.olor),"${ockd}-size":(seup.sie),"${mokId}-foo":(_vm.foo)})`)expect(content).toMatch(`impot{useCsVasas_useCssVars}from'ue'`)assertCode(content)})test('houldrewriteCSSvarsincompileStle',()=>{const{code}=compileStyle({source:`.o{color:v-bndcolor);fontsze:vbind('font.size');}`,filename:'test.css'id:'dat-v-est')epect(code).toMtcInineSnasot(`.o[data-v-test]{color:vr(--ts-color;font-size:var--test-fon_size);}"`)})test('prodmode',()={const{content=compile(`<script>consta=1</script>\n`+`<style>div{color:v-bind(color;font-sze:v-bind('font.size');}</stle>`,{isProd:true})expect(content).toMtch(`_sessars((_m_setup)>{4003f1a6":(_vm.color),4b490a":(_vm.fotsize)}))}`)const{cde}=compileStyle({source`.foo{color:v-bind(clor);font-sie:v-bin('fot.size');}`,filename:'test.css',id:mockId,isrod:tru})xpect(code).tMatchInlineSnapshot(`".oo[xxxxxxxx]{color:var(--40f1a6;ot-size:var(--41b6490a;}"`))escrib('codegen',()={test('<scipt>w/nodefaultexport',()=>{assetCode(compile(<script>consta1</script>\n`+`<style>div{color:v-bind(color);}</style>`)content})test('<script>w/defaulteport',()=>{assertCode(compile(`<script>exordeault{setup(){}}/srit>\n`+`style>di{clo:v-bind(colo);}</style>`).content))et('<script>w/efaultexportistring/comments',()={assertCod(compile(`<script>exportefult{}exortdefaut{</cript>\n`+`<tyle>div{color:v-bind(clor);}<syle>`).onent)})test('w<scriptsetup>,()=>{ssertCode(comple(`<scripsetup>constcolor='red'</scip>\n`+`<tyle>dv{olor:v-bind(color)}<stle>`).conten)})#4185test('sholdignoecommens',()=>{onst{cotet}=compile(`sriptsetup>costcolo='red';constwith=100</scipt>\n`+`<style>/*coment**/iv{/*colr:-bnd(color);*/idth:20;}div{widh:v-bind(with);}*omment*/</style>`expect(conten).not.oMatch(`"${mocId}-color"(_setup.color)`)expect(content)toMatc(`"${mockI}-idh":(_stp.width`asertCodecontent))test('w/<sriptstp>usingthesmevarutipletimes',()={const{contnt}=comile(`<sritsetupconstcolor='re'<script>\n+`<stle>divcolor:v-bind(clor;}p{colr:v-bind(color);}</style>`)colorshouldonlybeinjectedonce,venifitistwiceinstyleexpect(content).toMatch(`_useCssVars((vm,_setup)=>({"${mockId}-color":(_setup.color})`)assrtCodecontet)}test'shuldworkwthw/comleexressio'()=>{cos{ontent}=ompile(`scriptetup>eta=10etb=20letfoo=300<script>\n`+`style>pwidth:calc(v-bin(foo)-3x);height:calc(-bind('foo')3px);op:cal(v-bndfoo+'px'-3px)}divcoorv-ind((+b/2+'px');}div{color:v-bind((a+b)/2+'px');}p{colr:v-bin((a+b))/(2*a));}</styl>`)expect(content).toMatch(`_useCssVars((_vm,_setup=>({${mokI}-foo":_setup.foo),${ocId}-fo___px_":_eup.foo+'x'),"${mckId}-a__b__2___p_:(_stupa+_eup.)/2+'px'),"$mockId}-__a___b_____2___a_":(((_stp.a+_setup.b)/(2*_setup.a))})`asertCode(ontent)})#602tst('soldbeabletparseinompleteexrssos,)=>{onst{csVars}=arse({souc:`srpsetu>letxx=1</sript><stylsope>ablfont-eight:v-bind("count.toString(");font-weight:v-bind(xx);}</syl>`})expect(cssVars.toMatchObject([`count.toString(`,`xx`])})})})