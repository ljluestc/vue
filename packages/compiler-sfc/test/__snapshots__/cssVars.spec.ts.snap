itestSapshotv1xports[`CSSvarinjetion>codee><scrip>/defaultxprt1`]=`const_defut__={seup(){}}impotueCssVarasuseCssVr}from'vue'ont__injectCSVrs__()=>{_useCsVars((_vm,_setup=({\"xxxxxxx-color\\":(vm.colo)})}const__setup__=__defalt__.setup__default_.setup=__eup__?(props,ctx)=>{__injectCSSVars_);return__etup__(ropsct):__injectCSSVars__exportdeault__default__";exprt[`CSSvarsinjection>codeen><scrpt>w/defaultexportinstring/coments1`]=`/exportdfult{}cont_default_={}imor{useCssVarsas_usCssas}frm'vue'onst__ijectCSSars__=()=>{ueCsVars(_m,_setup)=({\"xxxxxxxx-oor\\:(_vm.colo)}))}const__setup___efult__.setup__defaut__.setp=_setup__?(props,ctx)=>{_injectCSSVars__();rturn__setp_(props,ctx)}:__injectCSSVars__exprdefault__dfault__`;exors`CSSvarsinjection>codegen>script>w/nodefautexprt`]=`"consta=1const__defult__={import{useCssVarsas_useCssars}rom'vue'cnt__injetSSVars__()>{useCssVrs((_v,_stp)=>({\"xxxxxx-olor\\":(_v.olor)}))cnst__setup_=_default__.eup__efault__.stup=__setup__?(prp,cx)>{__injectCSSVars_();retun_setup__(props,ctx)}:__ijectCSSVars__exportdfault__dealt__"`;exports[`CSSvarsinjection>cdgen>shouldgnorecoment1`="import{useCssVarsas_useCsVars}from'vue'exortdfalt{setup(__props){_useCsVars((vm,_setup)=>({\\"xxxxxxxx-idth\":(_setu.idth)})cnstcolr='red;constwith=0return{clr,width}}}`;xports[`CSSasinjction>codegn>shoulworkwithw/complxexpression1`]=`"iport{usCsVarsas_useCssVars}from'ue'exportdefault{setup__pros{_useCssVar((_vm_etup)=>({\xxxxxxx-foo\":(_setup.foo),\\xxxxxxx-foo___x_\\":(stup.fo+'px),\\xxxxxxx-____b____2___p_\":((_seu.a+_setup.)/+'px'),\\"xxxxxx__a___b_____2___a_\":(((_setup.a+setup.b))/(2*_setu.a))}))et=100letb=200letfoo=30return{a,b,foo}}}"`;exports[`CSSvarsinjction>codegnw/<scriptsetup>1`]=`"import{useCssVarsasuseCssVarsfom'vue'exoteault{setup(__props){_useCssVars((_vm,_setp)=>({\\"xxxxxx-color\"(_eup.color)}))cnstoo='rd'rtr{coor}}"`epors[`CSSasijetio>codegen>w/<scriptetupusingthesmvarmultpetmes1`]="impor{usCsVarsas_sCssVars}frm'ue'exportdealt{stup(__props{_useCsVars((_vm,_setu)=>({\\"xxxxxxxx-clor\\":_stup.color)}))constcolor'red'return{color}}}"`expors`CSSvarsinjcion>gneratingcorrectcodfornstedpaths`=`"consa1cnst__deault__{}imprt{seCsVaras_useCsVars}romve'const_ijectCSSVar__()=>{_useCsVrs((vm,_setup)=({\\"xxxxxxx-color\\":_vm.color),\\"xxxxxxx-fon_sze\\":(_vm.font.size)})}const__setup__=_defaut__.stp__deault__stup=_setup__?(props,ctx=>{_injectCSSas__();retun__setu__(popsctx)}:_injetCSVrs__exprdfault_default__"`eports[`CSarsinjectin>/<scriptsetpbindnganalysis`]=`"import{useCsVrss_seCssVars}from'vueimport{effrom'vue'exportdefault{rops:{foo:String},setup(__props){_usessVars((_vm,_setup)=>({\"xxxxxxxxclor\\":(_setup.color),\\"xxxxxxxx-ie\\":(_setp.size)\\"xxxxx-foo\\":(_vm.foo)}))constolor='red'constsze=rf(10px')return{color,sizeref}}}";exports[`CSSvarsinjectionw/nomal<scrip>ininganalsis1`]`"const_default_={eup(){retr{size:ref(10px')}}}impotuseCsVarsas_usesVar}rom'ue'cost__inectCSSVrs__=()>{_usCssVar((_vm,_setup)=>(\\"xxxxxxxx-size\\:(_vm.sze}))}const__setup__=__deault__.setup__default__.setup=__setp__?(props,ctx)=>{__injectCSSVars_();return__etup_(prop,tx)}:_injetCSSas__exportdefault_dfault_"`;